### PR TITLE
GH-351: Add context tester

### PR DESCRIPTION
**Pull request by lukasgisi**
Closes gh-351

### CODE REVIEW

File: frontend/src/app/tab-context-data/context-data-test/context-data-test.component.html [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    <form (ngSubmit)="testContext()" (reset)="reset()">
+        <mat-form-field class="form-full-width mr-1">
+            <mat-label>Name of Topic / Principal</mat-label>
+            <input name="testString" [(ngModel)]="testString" matInput>
+        </mat-form-field>
+        <button mat-flat-button color="primary" class="mr-1" [disabled]="(!testString) || pending">Test</button>
+        <button mat-flat-button type="reset">Reset</button>
+    </form>
+    @if (matchedTopicsContext || matchedPrincipalContext) {
+        <div class="mt-2">
```
Comment (lukasgisi) [inline frontend/src/app/tab-context-data/context-data-test/context-data-test.component.html:12]: Checked out the GUI quickly with Alex.
Group Topics / Principal in tabs for better visual clarity (e.g. there could be edge cases with lots of context for topic)

File: aggregator/src/test/java/io/spoud/kcc/aggregator/repository/ContextDataRepositoryTest.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                )
+        );
+
+        // when
+        List<ContextDataEntity> matchedContextData = contextDataRepository.testContext("test-topic-name");
+
+        // then
+        assertThat(matchedContextData)
+                .hasSize(3)
+                .map(ContextDataEntity::id).containsExactlyInAnyOrder("valid", "validFromYesterday", "validPrincipal");
```
Comment (arov00) [inline aggregator/src/test/java/io/spoud/kcc/aggregator/repository/ContextDataRepositoryTest.java:85]: Shouldn't it be in a specific order? (probably ordered by creation date)? We have to be sure that we can match the behavior that is tested here: https://github.com/spoud/kafka-cost-control/blob/master/aggregator/src/test/java/io/spoud/kcc/aggregator/stream/MetricEnricherTest.java#L323

File: aggregator/src/main/java/io/spoud/kcc/aggregator/repository/ContextDataRepository.java [hunk 3]
```diff
@@ -64,6 +72,27 @@ public List<ContextDataEntity> getContextObjects() {
...
+                matchedTopicList.stream(),
+                matchedPrincipalList.stream()
+        ).toList();
+    }
+
+    private static List<ContextDataEntity> findMatchedContextData(List<CachedContextDataManager.CachedContextData> cachedContextData, Metric metricToTest) {
+        return cachedContextData.stream()
+                // the matcher is only returned if it matches (and contextData is in correct time range, i.e. valid for "Instant.now()")
+                .filter(contextData -> contextData.getMatcher(metricToTest, Instant.now()).isPresent())
+                .map(contextData -> ContextDataEntity.fromAvro(contextData.getKey(), contextData.getContextData()))
```
Comment (arov00) [inline aggregator/src/main/java/io/spoud/kcc/aggregator/repository/ContextDataRepository.java:92]: I think we are missing a step here that substitutes placeholders with matched groups from the regex. Here's a related test from elsewhere in the code: https://github.com/spoud/kafka-cost-control/blob/master/aggregator/src/test/java/io/spoud/kcc/aggregator/stream/MetricEnricherTest.java#L535

File: aggregator/src/main/java/io/spoud/kcc/aggregator/repository/ContextDataRepository.java [hunk 4]
```diff
@@ -64,6 +72,27 @@ public List<ContextDataEntity> getContextObjects() {
         return list;
     }
 
+    public List<ContextDataEntity> testContext(String testString) {
+        List<CachedContextDataManager.CachedContextData> cachedContextData = cachedContextDataManager.getCachedContextData();
+        Metric dummyTopicMetric = new Metric(EntityType.TOPIC, testString);
+        Metric dummyPrincipalMetric = new Metric(EntityType.PRINCIPAL, testString);
+        List<ContextDataEntity> matchedTopicList = findMatchedContextData(cachedContextData, dummyTopicMetric);
```
Comment (arov00) [inline aggregator/src/main/java/io/spoud/kcc/aggregator/repository/ContextDataRepository.java:79]: I think that's a cool idea to use synthetic metrics. Could we perhaps even use a synthetic `TelegrafDataWrapper`? Then we could use its built-in `enrichWithContext` method to get the exact code running in the `testContext` that also currently assigns contexts in the main pipeline. Not sure if that would be easier or not.

**Review by arov00 — CHANGES_REQUESTED** [review]
Thanks for the PR. I have a couple of comments on the java part.

**Review by arov00 — APPROVED** [review]
Thanks for the adjustments, the backend part looks good to me now
